### PR TITLE
[frontend/backend] Adds account status and lockout functionality by expire date

### PR DESCRIPTION
**Pull request by ParamConstructor**
### Proposed changes

* Adds the following fields to the User schema (account_status and account_lock_after_date) 
* account_status is an ENUM of Active, Inactive, LockedTraining, or Locked
   - Active - normal account that would be active and able to login.
   - Inactive - is manually set by an administrator. There is a customizable warning message handled by (account_inactive_message in config/default.json) for user that attempt to login and are flagged to this account_status.
   - LockedTraining - is to support organization processes that allow for an account to be created in the background - but require a new user to take some sort of specific trainings to be allowed access. There is a customizable warning message handled by (account_locked_missing_training_message in config/default.json) for user that attempt to login and are flagged to this account_status.
   - Locked - is automatically tied to the optional account_lockout_after_date field. There is a customizable warning message handled by (account_locked_message in config/default.json) for user that attempt to login and are flagged to this account_status.
* account_lockout_after_date (optional field) - if populated will mark the User account Locked after the specific date. Users attempting to login after this date will receive the configured account_locked_message 
* NOTE: Changing to the account_status value on an account will result in termination of any "active" sessions. The point being, should you as an Admin be told to Lock or mark an account Inactive, you would not want - if that user were currently logged in - a session to persist for them after that flag was set. So, sessions are terminated for a user upon account_status change.

MIGRATION NOTE:

Upon startup of OpenCTI after these changes have been adopted:
* A function (initUserAccountStatus) has been added to `opencti-platform/opencti-graphql/src/config/providers.js`. This function will "repair" / "seed" any User accounts that are missing a valid ENUM status for account_status to Active (default for default_initialize_account_status from conf/default.json)
   - Running of this feature can be disabled by setting initialize_account_status in conf/default.json to false for future startups - but needs to be true on initial migration or all existing accounts previous to this capability will have a NULL account_status resulting in lockout.
* The function (initUserAccountStatus) is called by `opencti-platform/opencti-graphql/src/initialization.js` at system startup
  - The internal Administrator account will always be set at startup to Active

### Related issues

* Automation support for the LockedTraining status is planned for a possible additional future PR.

### Checklist

- [ X ] I consider the submitted work as finished
- [ X ] I tested the code for its functionality
- [ X ] I wrote test cases for the relevant uses case
-    Manually tested. There are Selenium automated tests that can be submitted as part of a Selenium GUI auto test future PR.
- [ ] I added/update the relevant documentation (either on github or on notion)
- [ X ] Where necessary I refactored code to improve the overall quality

### Further comments

None.

### CODE REVIEW

**Comment by ParamConstructor** [comment]
@richard-julien - rebased to [cd9069b](https://github.com/OpenCTI-Platform/opencti/commit/cd9069b91fba5e3c467dc6cfe8420283d729b7b8) - (i.e. 5.8.5 + 1 commit)

**Comment by SouadHadjiat** [comment]
Hello @Bonsai8863 @ParamConstructor  ! We've seen that the tests are failing on this branch, since `account_status` has been set as a mandatory property when creating a User (in `UserAddInput` graphql definition), and we create users in `testQuery.ts` without this property.
I'm wondering, why is it mandatory to set `account_status` on creation ? Since the feature is optional with `initialize_account_status` set to false.

**Comment by Bonsai8863** [comment]
> Hello @Bonsai8863 @ParamConstructor ! We've seen that the tests are failing on this branch, since `account_status` has been set as a mandatory property when creating a User (in `UserAddInput` graphql definition), and we create users in `testQuery.ts` without this property. I'm wondering, why is it mandatory to set `account_status` on creation ? Since the feature is optional with `initialize_account_status` set to false.

Hi @SouadHadjiat,
I made the `account_status` property in `UserAddInput` optional, but set its default to 'Active'. For clarification, `initialize_account_status` does not make the `account_status` property optional; it makes migration of existing users easier by initializing their account statuses.
I also made the `password` and `user_email` properties in `UserAddInput` required, as the `userAdd` resolver fails if these properties are not provided in its input.
I hope this helps and feel to ask us for any additional clarification.

CC: @ParamConstructor

**Comment by ParamConstructor** [comment]
Branch rebased to 5.9.4 (a3067f0)

**Comment by richard-julien** [comment]
Hi @ParamConstructor , looks like you continue to do some massive change / approach change at the same time i push a new approach for the feature. So i will stop to work on it and let you finished what you imagined.
You can look into my last commit and schedule a meeting with me as soon as possible in order to have a better synchronization.

Thanks

**Comment by ParamConstructor** [comment]
@richard-julien - We have been "done" with the things we were doing with this PR. We have just been pushing updates to rebase it to master weekly - as you guy update. Feel free to make modifications you require. I see you have committed some to streamline / make it better. We shouldn't have anything "new" we are adding to it. Thank you for the updates to make it more inline with your development strategy.

**Comment by richard-julien** [comment]
> @richard-julien - We have been "done" with the things we were doing with this PR. We have just been pushing updates to rebase it to master weekly - as you guy update. Feel free to make modifications you require. I see you have committed some to streamline / make it better. We shouldn't have anything "new" we are adding to it. Thank you for the updates to make it more inline with your development strategy.

Hi @ParamConstructor , perfect thanks. I just finished and committed my work.
So the major changes if you want to talk a look and do some testing.
- Replace init by a migration
- Replace static status and message by configuration
- Add auto expiration in expired manager
- Move extra management front frontend to backend (like session kill)

**Comment by ParamConstructor** [comment]
@richard-julien - Thank you for the updates. We will attempt to follow this style/function placement in future PRs, if needed. Example - movement of init code into a migration versus how we had done it to just check on startup with a flag. I believe all the changes made - leave the functionality like intend. I believe they also still meet the requirements of this feature request (https://github.com/OpenCTI-Platform/opencti/issues/2254). We will give run it through some testing tomorrow - but believe it should all function as intended and be ready for joining the mainline/master - when you wish to approve it. Thank you for the updates and review.

**Comment by ParamConstructor** [comment]
@richard-julien - Tested it. Works great. Thank you. Did add "Expired" in "locked_account_statuses" with a customizable message. Otherwise - I believe it is ready for merge when you are ready to accept it.

File: opencti-platform/opencti-front/src/private/components/settings/users/User.tsx [hunk 1]
```diff
@@ -326,6 +328,12 @@ const User: FunctionComponent<UserProps> = ({ userData, refetch }) => {
     .sort(
       (a: Session, b: Session) => timestamp(a?.created) - timestamp(b?.created),
     );
+
+  let accountExpireDate = fldt(user.account_lock_after_date);
+  if (accountExpireDate === '-') {
+    accountExpireDate = 'None';
```
Comment (Archidoit) [inline opencti-platform/opencti-front/src/private/components/settings/users/User.tsx]: We decided that empty fields should be displayed as '-' (and not 'None') (to be homogeneous with the rest of the platform).

File: opencti-platform/opencti-graphql/src/migrations/1691652922595-init-account-status.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { elRawUpdateByQuery } from '../database/engine';
+import { READ_INDEX_INTERNAL_OBJECTS } from '../database/utils';
+import { DatabaseError } from '../config/errors';
+import { DEFAULT_ACCOUNT_STATUS } from '../config/conf';
+
+export const up = async (next) => {
+  const updateUserQuery = {
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/migrations/1691652922595-init-account-status.js]: can you log a message at the beginning and at the end of your migration (or use elUpdateByQueryForMigration that log those messages)?

File: opencti-platform/opencti-graphql/config/schema/opencti.graphql [hunk 3]
```diff
@@ -992,6 +992,10 @@ type Provider {
   strategy: String
   provider: String
 }
+type UserStatus {
+  status: String!
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/config/schema/opencti.graphql:1014]: I wonder if it would be better if status is an enum. What do you think?

**Comment by Archidoit** [comment]
Can you resolve the conflicts you have with the main branch? Thanks

File: opencti-platform/opencti-graphql/config/schema/opencti.graphql (hunk 3 above)
Comment (ParamConstructor) [reply opencti-platform/opencti-graphql/config/schema/opencti.graphql:1014]: @Archidoit - they were an ENUM on the initial submission but in this commit @richard-julien swapped the behavior. (https://github.com/OpenCTI-Platform/opencti/pull/3303/commits/ccd802daa29109878c3438cb4c2d08c88700a275)

Comment (Archidoit) [reply opencti-platform/opencti-graphql/config/schema/opencti.graphql:1014]: okay, let's keep a string then

**Comment by Archidoit** [comment]
Can you also change the migration number (so that the migration file is the latest to run after merging on the main branch)?

File: opencti-platform/opencti-front/src/private/components/settings/users/User.tsx (hunk 1 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/users/User.tsx]: image

**Comment by Archidoit** [comment]
What about displaying the account status in a special box (design)?
image

You can use the ItemStatus component and get inspiration from the capture below for instance
image

**Comment by Archidoit** [comment]
Behavior tested and works well :)

File: opencti-platform/opencti-front/src/private/components/settings/users/User.tsx (hunk 1 above)
Comment (ParamConstructor) [reply opencti-platform/opencti-front/src/private/components/settings/users/User.tsx]: @Archidoit - Committing this fix now.

File: opencti-platform/opencti-graphql/src/migrations/1691652922595-init-account-status.js (hunk 2 above)
Comment (ParamConstructor) [reply opencti-platform/opencti-graphql/src/migrations/1691652922595-init-account-status.js]: @Archidoit - Committing this fix now and moved filename to new epoch time

**Comment by ParamConstructor** [comment]
@Archidoit - Something like this maybe? (just hardcoded onto page to get a screen grab of all options - obviously only one would be displayed)
image


--> @ParamConstructor yes that's perfect.

**Comment by ParamConstructor** [comment]
@Archidoit - Status Display change is pushed - it is building now. Thank you for the feedback and peer review.

File: opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx [hunk 4]
```diff
@@ -264,6 +258,28 @@ class UserCreation extends Component {
...
+                  <Field
+                    component={SelectField}
+                    variant="standard"
+                    name="account_status"
+                    label={t('Account Status')}
+                    fullWidth={true}
+                    containerstyle={fieldSpacingContainerStyle}
+                    >
+                    {settings.platform_user_statuses.map((s) => {
+                      return <MenuItem key={s.status} value={s.status}>{s.status}</MenuItem>;
```
Comment (Archidoit) [inline opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx]: What about translating account statuses?

File: opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx [hunk 5]
```diff
@@ -192,10 +184,12 @@ class UserCreation extends Component {
                 password: '',
                 confirmation: '',
                 objectOrganization: [],
+                account_status: 'Active',
```
Comment (Archidoit) [inline opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx:187]: translation?

File: opencti-platform/opencti-front/src/private/components/settings/users/User.tsx [hunk 6]
```diff
@@ -405,6 +411,22 @@ const User: FunctionComponent<UserProps> = ({ userData, refetch }) => {
                 </Typography>
                 {user.lastname || '-'}
               </Grid>
+              <Grid item={true} xs={6}>
+                  <Typography variant="h3" gutterBottom={true}>
+                    {t('Account Status')}
+                  </Typography>
+                  <ItemAccountStatus
+                    account_status={user.account_status}
+                    label={t(user.account_status || 'Unknown')}
```
Comment (Archidoit) [inline opencti-platform/opencti-front/src/private/components/settings/users/User.tsx:420]: missing account statuses translations in Localization.js

File: opencti-platform/opencti-front/src/private/components/settings/users/UserEditionOverview.tsx [hunk 7]
```diff
@@ -266,20 +287,35 @@ UserEditionOverviewComponentProps
...
+              fullWidth={true}
+              containerstyle={fieldSpacingContainerStyle}
+              onFocus={handleChangeFocus}
+              onChange={handleSubmitField}
+              helperText={
+                <SubscriptionFocus context={context} fieldName="account_status" />
+              }
+            >
+              {settings.platform_user_statuses.map((s) => {
+                return <MenuItem key={s.status} value={s.status}>{s.status}</MenuItem>;
```
Comment (Archidoit) [inline opencti-platform/opencti-front/src/private/components/settings/users/UserEditionOverview.tsx]: same

File: opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx (hunk 4 above)
Comment (ParamConstructor) [reply opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx]: @Archidoit - give me a little bit - I will push a commit with the translations fixed - sorry.

File: opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx (hunk 5 above)
Comment (ParamConstructor) [reply opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx:187]: @Archidoit - Stored internally as 'Active' regardless of selected language selected. On display, the MenuItem display value will translate this with the correction you identified.

File: opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx (hunk 4 above)
Comment (ParamConstructor) [reply opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx]: @Archidoit - Corrected.

File: opencti-platform/opencti-front/src/private/components/settings/users/User.tsx (hunk 6 above)
Comment (ParamConstructor) [reply opencti-platform/opencti-front/src/private/components/settings/users/User.tsx:420]: @Archidoit - Corrected. Also, did find a few existing 'unknown' translations that were typo'ed with 'unkown', so corrected those.

File: opencti-platform/opencti-front/src/private/components/settings/users/UserEditionOverview.tsx (hunk 7 above)
Comment (ParamConstructor) [reply opencti-platform/opencti-front/src/private/components/settings/users/UserEditionOverview.tsx]: @Archidoit - Corrected.

File: opencti-platform/opencti-front/src/private/components/settings/users/User.tsx (hunk 6 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/users/User.tsx:420]: thanks for the changes!

File: opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx (hunk 5 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/users/UserCreation.jsx:187]: yes you're right!

**Comment by Archidoit** [comment]
Thanks for all the changes! I'm going to do a final review. Can you just resolve the conflicts you have with the main branch before? Thanks

**Comment by ParamConstructor** [comment]
@Archidoit  - Conflicts resolved, rebased, committed/pushed, and building now in the drone pipeline.

**Comment by Archidoit** [comment]
Tested and works well 👍 
Waiting for some minors modifications of @richard-julien before merging

**Pull request merged**